### ISSUE TITLE
Move Case to Postgres

### ISSUE THREAD

**Issue by pixiwyn** [ustaxcourt/ef-cms#8836]
branch off: origin/10391-dynamo-migrations-kysely-aurora-rds-tokens (till merged in)

DynamoDB Entities to Replace:
- case

OpenSearch Indexes to Keep:
- efcms-case (used to search cases)

convert the functions in web-api/src/persistence/dynamo/cases, and web-api/src/persistence/dynamo/caseWorksheet to postgres kysely functions.

DynamoDB Functions to Move (unit tests can be removed):
<img width="242" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMU1BQnc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--9fc085732721d4c30822436d458c8ec20218192a/2024-09-25_11-24-03.png" alt="2024-09-25_11-24-03.png" />

**Comment by Mwindo** [comment]
Since we'll need to test cases thoroughly anyway, we will also try to implement a CaseFactory as part of this ticket. See https://github.com/ustaxcourt/ef-cms/pull/5557.

**Comment by Mwindo** [comment]
We have moved over all of the seed data and have the app "running" locally (imperfectly). The next step is to get the app fully working locally and refactor. After that, it would be good to look at real data and eliminate unnecessary fields to simplify our tables (or include any fields we are missing).

**Comment by Mwindo** [comment]
We have the app working locally after fixing various issues. Current work:

- Investigating how to get database updates indexed in Open Search. Right now, the current thought seems to be something like: `Persistence method called --> Change Data Capture tool (probably AWS DMS or Data Activity Stream, although database plugins of various sorts are possible too) --> Kinesis --> Lambda --> Open Search`
- Deciding if it would be worth taking a detour to implement performance monitoring (which can be tweaked but is otherwise more or less ready) _before_ deploying the change to migrate `Case` over to postgres.
- Fixing various TODOs and refining the code. Case touches everything, so we want to get things right.

**Comment by Mwindo** [comment]
A few scattered notes:
 
- We have decided that the logs in Open Search should be enough and that we do not need the performance monitoring for now.
- We explored Data Activity Stream. Unfortunately, Data Activity Stream only works with provisioned Aurora, not serverless Aurora.
- I have been exploring Data Migration Service. The issue here is that the service is near real-time (NRT), which means we could experience delays of minutes. This is not an acceptable latency for certain tasks (e.g., searching by case docket number).
- I have briefly explored other options. In general, and unsurprisingly, provisioned Aurora seems to play more nicely with CDC tooling (e.g., Debezium). I am still looking into these options.
- A "last-resort" option is to move into Postgres those Open Search persistence queries requiring low latency and then using DMS for the rest. This is not a terrible option considering we want to move these persistence queries into Postgres anyway.

**Comment by Mwindo** [comment]
After some further research, my findings are basically the following:

Serverless Postgres, while great for scaling up and down and for not having to worry about dependencies, etc., is not ideal for low-latency CDC. Most tools work better, and often exclusively, with provisioned Aurora. As far as I can tell, out options include:
- DMS: This is the general approach people seem to use, but it isn't meant for realish-time data ETL.
- Postgres triggers that invoke a lambda to stream data into Kinesis: fast, but less robust (what if the lambda fails?)
- Postgres triggers with pg_notify and a dedicated listener: a worse version of the above since we would need to have a dedicated listener
- Doing stuff in application logic. The best example is probably something like making a wrapper for `getDbWriter` that writes to postgres and kicks off a kinesis task. This is fast and keeps the two in sync, but we have to handle failure logic to kinesis.
- Some postgres plugin (?) for CDC that plays nicely with serverless setup

Because of this, I discussed with Jim, and we are doing an experiment to see how difficult it is to just rewrite our Case open search queries in postgres. If this is not too difficult, we can probably just port most things over. The only thing that _needs_ to be in Open Search, as far as I know, is Docket Entry text. If that's the case, the worries above are basically moot: latency is less of an issue, and we don't need to worry as much about having a highly scalable solution.

**Comment by Mwindo** [comment]
General updates:

- Most Case queries that used OpenSearch now use Postgres; however, `caseAdvancedSearch`, `casePublicSearch` (which is basically the same minus sealed cases), and `advancedDocumentSearch` proved too difficult to replicate. OpenSearch is a better tool for these scenarios. We have explored various ways of getting data into OpenSearch given the constraint that we need seconds latency between time of update to DB and time of indexing in OpenSearch. We have tentatively settled on the following: call to getDbWriter --> write to an SQS queue (in application code) --> trigger lambda --> OpenSearch. The downsides are 1) slightly more complicated application code, 2) changes that are made _outside_ of DAWSON will not automatically update OpenSearch, and 3) potentially more complex transactions in multithreaded scenarios. The advantages are 1) no separate infrastructure (e.g., a FarGate instance), 2) lower costs, and 3) speed. We have made good headway on this (e.g., we have a proof of concept working locally), also investigating how transactions will work in the future, but we need to set up some terraform in a deployed environment to fully test.
- We have fixed many tests (and any errors they reveal) and are still fixing them.
- We have a few TODOs to clean up, and then we need to test everything in a deployed environment.

This ticket has been, unsurprisingly, complex: the backend is very heavily coupled to DynamoDB.

**Comment by ttlenard** [comment]
Testing Feedback:
@Mwindo @pixiwyn 

- [x] The Custom Case Report is currently not displaying filtered data correctly.  I put in this search (Paper filed, start date of 1/1/25), and I am getting very old case data.  If you put in a start AND end date, then the results populate correctly.

![Image](https://github.com/user-attachments/assets/44b09ad8-a08e-4184-b353-72ce10904193)

**Comment by Mwindo** [comment]
The above issue is resolved. Here are some ongoing issues we are resolving:

- [x] 1. The Cold Case report can time out. (Sometimes it is snappy, but this is just when Postgres happens to have the relevant data in its cache.) For now, we are going to revert this report to OpenSearch. There are ways to improve the efficiency of the underlying query in Postgres, but that's better to tackle later as needed.
- [x] 2. The app times out when attempting to load the case details for docket numbers 15631-22 and 16017-21. (This also means their associated trial sessions page times out.) We think this is a memory issue: essentially, we are pulling too much data at once in Postgres. We think we have a fix for this.
- [x] 3. There is an issue with Case Deadlines and https://github.com/flexion/ef-cms/issues/10576. We are debugging with John's help.
- [x] 4. Indexing into OpenSearch case message size issue:

<img width="1333" alt="Image" src="https://github.com/user-attachments/assets/f04d39a1-8b1b-41f6-83b1-bd9e1fd0e277" />

- [x] 5. Flaky tests due to race conditions (trying to create a case with the same docket number). This seems to be resolved. We will monitor.
- [x] 6. Custom case report sometimes times out.

**Comment by Mwindo** [comment]
I think we have fixes ready for 4 and 6 above, but they are not yet in test. 3 is ongoing, but perhaps not critical now that John has refactored to use an asynchronous endpoint.

**Comment by ttlenard** [comment]
Testing Feedback:

@Mwindo @pixiwyn 

We might be able to remove the following fields from the Case table. 

1) Currently, all of the following only have NULL values in the Postgres table.  Please confirm.
- [x] can_allow_document_service
- [x] can_allow_printable_docket_record
- [x] judge_user_id
- [x] can_doj_practitioners_represent_party


2) Looks like the hearings column is only ever [ ] or NULL.  Should we remove this?
- [x] hearings


3) Looks like we are only sometimes populating the docket_entries column. Only around 2700 cases had any data in the docket_entries column, the rest were NULL.  Does this column even need to be in the case table? 
- [x] docket_entries

**Comment by ttlenard** [comment]
Testing Feedback:

@Mwindo @pixiwyn

- [x] If a Petitions clerk is QCing a Petition and adds in an IRS notice date, and then a statistic, clicks review petition, and THEN edits the IRS notice information to remove the IRS notice date and delete the statistic data, upon clicking on Review petition, the UI shows that statistic data still on the screen (but the IRS notice date does not display anymore on the UI). However, in Postgres, the IRS notice date and the statistics are still saved in the table.  When you go back in to edit the IRS notice information on the UI, the date is displayed again, along with the statistic data.
- [x] I notice that the Automatic blocked reason and Automatic blocked date are retained in Postgres after you remove the Pending item or deadline that caused the automatic block. Should these be cleared out?
- [x] Can we also clear out any data that is retained in for blocked reason and blocked date if there is no block?

**Comment by Mwindo** [comment]
@ttlenard: I believe all of the feedback thus far has been addressed. Commenting on a few highlights:

1) Those six fields you pointed out indeed look like they do not need to be stored. They are no longer stored.
2) Statistics are cleared properly in the QC process you discovered.
3) Blocked and automatic blocked reason/date are cleared properly.
4) The custom case report is not timing out for me on the East Coast. (Previously, for instance, selecting the status "General Docket - at issue ready for trial" led to a timeout.)
5) I haven't found a case that times out. Some bigger ones I have tested: 22447-16, 21959-16, 16017-21, 16017-22, 15631-22.

General note: the database caches recent queries. This means if you see a timeout when trying to load a case, a report, a trial sessions page, etc., then try again and don't see a timeout, you should not think, "Huh, must have been a fluke?" and instead think, "Hmm ... seems like a timeout issue, better let the team know."

**Comment by ttlenard** [comment]
Testing Feedback

@Mwindo @pixiwyn 

On the dw_petitioner_on_case table.  I'm seeing a few things that are weird. For example:

- [x] has_consented_to_electronic_service is NULL for every single record, but it did get populated to TRUE on on a brand new case that I just created. Shouldn't this be set to TRUE for cases already in DAWSON in which that checkbox was checked?

- [x] has_electronic_service is NULL for every single record. Not sure why?

- [x] sealed_and_unavailable is FALSE for every single record - why?

**Comment by ttlenard** [comment]
Testing Feedback:

@Mwindo @pixiwyn 

- [x] I'm having a hard time getting the additional_name field to be populated.  What logic is in place to get petitioner contact data to populate in this field?

**Comment by ttlenard** [comment]
Testing Feedback:

- [x]  I notice that the "use_same_as_primary" column is in the "dw_case" table.  Would this be more appropriate in the "dw_petitioner_on_case" table instead? NOTE: This field is set to true ONLY when a Court User is inputting a paper petition and they check the box to use the same address as the primary petitioner.  This is currently NOT getting set to TRUE when an electronic petitioner checks the box to use the same mailing address as above.  I assume it should be getting set to True for this workflow as well?

**Comment by ttlenard** [comment]
@Mwindo @pixiwyn 

Testing Feedback:

- [x] When a case is eFiled by a Petitioner, the case status is New.  This data is not being reflected in the Case History table, or on the History tab in the case.  There should be an entry for New status. In this scenario, it should track that the Petitioner

This screen grab shows no data in Postgres

![Image](https://github.com/user-attachments/assets/3a27b01e-96ed-4be4-af79-bda8fb98e004)

This screen grab shows no data in the case history table in the UI

![Image](https://github.com/user-attachments/assets/33e20903-ad1c-4b28-91ae-3f659732b016)


You can see that the New status used to be tracked here. Here is an example from the internal user guide:

![Image](https://github.com/user-attachments/assets/7ef8c1d9-e01f-48ce-b502-299f44a6565c)

**Comment by Mwindo** [comment]
@ttlenard I will double check, but I believe this is the current behavior:

1. On creating a case, there is no case history. The most recent case status is on the case record/table (sk of case|whatever in Dynamo, dw_case in Postgres) itself.
2. On changing the status from X to Y, the old status X is then appended to the history, and Y is saved on case. And so on.

So I believe that you need to change the status on the case to have the previous status reflected in the status history.

**Comment by ttlenard** [comment]
@Mwindo I'm going to have to disagree with your last comment :)

See these cases:

I filed this case on test as a petitioner, but it isn't showing the New status, it is only displaying the status when the status got updated:

![Image](https://github.com/user-attachments/assets/b9e44d9e-5dce-4462-9448-1219296e42e2)

Refer to this case that was existing in the system. Notice that it displays the New status, and the "Changed by" displays "Petitioner". This is what I was expecting to see on the above case.

![Image](https://github.com/user-attachments/assets/8a472453-0ea9-4e9b-a995-d0503660fa05)

As another comparison, see this example of an existing case in the system, where a Petitions clerk input a paper filed case.  For this workflow, we display the petitions clerks name, with the status of New.
 
![Image](https://github.com/user-attachments/assets/b2cba6f5-7552-46ce-9d26-c0ee2ce0653e)

**Comment by Mwindo** [comment]
@ttlenard Confirmed! Thank you for catching this. I will address this later today and add a test as well.

**Comment by Mwindo** [comment]
Also, responding to a few comments here that I missed:

> I'm having a hard time getting the additional_name field to be populated. What logic is in place to get petitioner contact data to populate in this field?

We figured this one out together, but including it here for posterity! This happens when a case is updated [here](https://github.com/ustaxcourt/ef-cms/blob/cb7d52a1bfc201e2b6dd420bd7164931e26e7db2/shared/src/business/entities/cases/Case.ts#L2369-L2410). On first creating a case, it will not be populated.

> I notice that the "use_same_as_primary" column is in the "dw_case" table. Would this be more appropriate in the "dw_petitioner_on_case" table instead? NOTE: This field is set to true ONLY when a Court User is inputting a paper petition and they check the box to use the same address as the primary petitioner. This is currently NOT getting set to TRUE when an electronic petitioner checks the box to use the same mailing address as above. I assume it should be getting set to True for this workflow as well?

Looking into this.

**Comment by ttlenard** [comment]
Confirmed in Dynamo that the "use_same_as_primary" is set to true only when a Court user inputs a paper petition and they check the box to use the same address as primary petitioner. This field is not present in Dynamo for eFilers that check the box to use the same address as their spouse.

No action needed at this time for this field, but I would probably like to see this field stored with the dw_petitioner_on_case table in the future.  We'll add this to the list of things for later.

**Comment by ttlenard** [comment]
Testing Feedback:

- [x] Case Services Supervisor marks a case as high priority. When they do this, they have to add a reason.  After removing the case from being high priority, it is currently not removing the data in the high_priority_reason field.

Here is a screen grab from postgres table. Notice that the reason is still there, even though the high_priority is false. Should this be getting cleared out if the case is no longer marked as high priority?

![Image](https://github.com/user-attachments/assets/aeeb2af3-8267-4781-9bcc-80884175126f)

**Comment by Mwindo** [comment]
> Case Services Supervisor marks a case as high priority. When they do this, they have to add a reason. After removing the case from being high priority, it is currently not removing the data in the high_priority_reason field.

Got it. I know the issue and will fix this in the next PR I open as well.

**Comment by Mwindo** [comment]
@ttlenard, I've pushed fixes for the high priority reason and case status history issues above and tested briefly.

**Comment by ttlenard** [comment]
Testing Feedback:

- [x] I just got a 504 on test when a chambers user was attempting to open this trial session copy: https://app.test.ef-cms.ustaxcourt.gov/trial-session-working-copy/9b2e3a65-8625-4349-bfc4-1804dd51e8c9. I didn't have the console open at the time, but it was able to capture this:

![Image](https://github.com/user-attachments/assets/329429cc-1254-4331-bd00-3db73399413f)

After chatting with Chris, this seems to be a performance issue caused by Postgres, and he is working on fixing it. Thanks Chris!

**Comment by Mwindo** [comment]
^ @zachrog Is on this with me!

The issue is [here](https://github.com/ustaxcourt/ef-cms/blob/cb7d52a1bfc201e2b6dd420bd7164931e26e7db2/web-api/src/persistence/dynamo/trialSessions/getCalendaredCasesForTrialSession.ts#L24-L32). We are making several database queries for every case on the trial session. If there are many cases on the trial session, this can take time. We are working on refactoring to optimize.

**Comment by ttlenard** [comment]
Testing Feedback.  

- [ ] Both Chris H. and I ran into some issues with the page displaying information correctly, and after a conversation with Zach and Chris, they think it may have to do with East Coast/West Coast issues.  Here are the specifics:

Admissions clerk added correspondence to a case. After upload, this was the display. Notice that there is the confirmation message indicating that it was added, but no file.

![Image](https://github.com/user-attachments/assets/0ffd1159-0ef7-4541-8e6c-bfbfbeb97659)

Admissions clerk updated Petitioner contact info, the result shows the confirmation message, but no petitioner info.

![Image](https://github.com/user-attachments/assets/66d3de32-2513-4fce-b5d6-3fd4d8016ac6)

**Comment by ttlenard** [comment]
Testing feedback

- [x]  The ordering of the Consolidated Cases on the Case information page is no longer displaying in order by latest to newest docket number when the following workflow occurs. I double checked on staging and the order of the docket numbers is retained on Staging. This seems to only be happening on Test.

1. Case Services Supervisor accesses a member case 
2. Click on Case information tab
3. Click on the remove cases link on the consolidated cases card
4. Remove the lead case docket number (at this point, the list order is correct)
5. After the case is removed, you will get a green confirmation banner indicating that the "selected cases consolidated", but notice that the listing of docket numbers on the consolidated cases card are now out of order.
6. If you refresh the page, the list updates.
7. The same ordering of docket numbers issue occurs if you add the case that you removed back in. The list order is off.

In this screen grab, I had just removed the lead case docket number 1356-23. Note that the new lead case 1357-23 is not displayed at the top of the list.

![Image](https://github.com/user-attachments/assets/c234883d-d845-49c3-80ba-6c20ed365cfc)

**Comment by Mwindo** [comment]
Thank you @ttlenard! Our current next steps:

- Fix the performance issue you noticed. Preemptively fix similar performance issues that could occur elsewhere.
- Fix the east-west coast issue you and Chris ran into, and preemptively fix others that could occur elsewhere.
- Fix the ordering of consolidated cases.

**Comment by ttlenard** [comment]
@zachrog @Mwindo 

- [x] Looks like I am still running into an east/west issue when adding correspondence.  I'm either getting the same results as what I reported above, or if there is more than one correspondence file on a case, it looks like this. In this screen grab, it is only showing the correspondence file of "test", but I had just added another correspondence document to this case and I get this:

![Image](https://github.com/user-attachments/assets/12d6f8ff-3804-4725-a880-69c8f35baec9)

**Comment by zachrog** [comment]
- [x] Fix east/west issue when adding a consolidated case or removing a consolidated case from a group

**Comment by Mwindo** [comment]
We have run into odd data inconsistencies on test. (Thanks @ttlenard!) After investigating, our suspicion is that the issue lies in our lack of transactions: if we are awaiting updates to hundreds of cases and one fails, we can interrupt other processes and cause data to be in an inconsistent state. This issue is now more salient than in the past because petitioners and statistics are no longer stored on the case record itself. (When they were stored on the case record itself, you could not save petitioners and statistics separately from a case. Now, because we do, an interruption can mean the case has been updated, but the statistics have not, etc.)

We will be looking into adding the minimum necessary transaction support to prevent this. So our goal will not be to fully address https://github.com/flexion/ef-cms/issues/10467.

To recap, here is our current overall TODO:

- [x] Minimum necessary transaction support to prevent inconsistent data
- [x] The sort order of consolidated cases when removing one from the group
- [x] The east-west issue when adding/removing a consolidated case from a group
- [x] The upload-correspondence east-west issue, along with other possible east-west issues not related to those we already fixed

**Comment by Mwindo** [comment]
To fix east-west issues entirely, we discussed with the team the idea of doing away with us-west-1 altogether (except for some backup data). In this case, all requests would go to the us-east-1 lambda, which would read from a single database. (This would mean no east-west replication lag since we are only ever reading/writing from the east.)

Benefits of this:
- Simpler infrastructure and related code
- Cheaper
- Faster for west users on all but the most trivial workflows
Potential cons:
- No longer spreading out DB load. But, looking at AWS metrics like ACU utilization, it looks like this shouldn't be an issue.
- No "ready-to-go" read replica in the west

As a whole, we thought the pros heavily outweighed the cons. We will proceed with this approach, and this will get rid of the east-west issues now and in the future.

N.B.: We never read from us-west-1 with Dynamo. We were always reading from us-east-1.

**Comment by ttlenard** [comment]
- [x] Removing/adding cases to a consolidated group is timing out.

I was following the steps from the issue reported above and I ran into a slightly different issue.

1. Case Services Supervisor accesses a member case (I used 23186-16)
2. Click on Case information tab
3. Click on the remove cases link on the consolidated cases card
4. Remove the lead case docket number (I removed case 12027-16). At this point, the list order is correct
5. Receive this error. Notice that it would appear that the lead case is still in the group.

![Image](https://github.com/user-attachments/assets/0fbb1b6e-b6a1-4f2d-8993-834e8bb03a2c)

6. Refresh the page and go back to the case info tab, notice that the sort order is correct, and the lead case from the steps above IS actually removed from the group.
7. Click on the Add Cases link
8. Type in the lead case from before (12027-16)
9. Receive 400 error.

![Image](https://github.com/user-attachments/assets/4bb1502a-6ade-4843-a9d3-2cee08d040ab)

**Comment by ttlenard** [comment]
- [x] Might have another timeout issue with filing multiple documents across a consolidated group of cases. As a Private Practitioner, I just attempted to file a motion for leave to file + 5 supporting documents, along with another motion with 5 supporting documents in case 12722-18 and across the consolidated group that I found in test that had 24 cases.  Got a timeout. Interestingly, this was the user experience. 

![Image](https://github.com/user-attachments/assets/c892d9f9-45eb-4464-ba94-427fc51d53b6)

**Comment by ttlenard** [comment]
UPDATE: CREATED A SEPARATE BUG FOR THIS ustaxcourt/ef-cms#8971

- [x] I found an error with changing a Judge on an XL trial session.  

- Log in as a Petitions clerk or CSS
- Navigate to an open XL trial session (200 cases)
- Click on the Edit link on the trial session page
- Change the Trial Judge
- Click Save
- You will get the spinning wheel for over 15 minutes
- Check the logs and see this error:

![Image](https://github.com/user-attachments/assets/db0969e8-c1ce-47eb-8ebb-c716911f3832)

I also see a 400 error:

![Image](https://github.com/user-attachments/assets/20215b1b-4d24-4e02-872d-becfa9166b32)

Note:
- I also got this same error when attempting to change the judge a second time, and then I got it again when attempting to change a different XL trial session from in person to remote. Same error also received on a "normal" trial session with 100 cases.

![Image](https://github.com/user-attachments/assets/544aa787-c731-44f1-9aba-72a819f75ab6)

**Comment by Mwindo** [comment]
@ttlenard As we discussed after you posted this (just trying to clean up our paper trail here), we think the above error is not related to Postgres, but it should be investigated before release.

@zachrog and I are making a few smaller changes now to improve the code and app efficiency, which we will push soon, but we think everything except the above error has finally been addressed.

**Comment by ttlenard** [comment]
Update: Created bug ustaxcourt/ef-cms#8971 to address the latest comment re: "Unable to generate pdf" error.

**Comment by Mwindo** [comment]
@ttlenard After addressing ustaxcourt/ef-cms-flexion#10658 (a bizarre bug having to do with environment variables passed into puppeteer) and ustaxcourt/ef-cms#8973 (we made things faster, with the result that we had too many lambda connections open at once!), 10502 is back in the "we have addressed everything we know of" state!
